CenterForOpenScience/ember-preprints
# [rfr] Support cookie login and OSF integration

## Requires
https://github.com/CenterForOpenScience/osf.io/tree/feature/ember-preprints
https://github.com/CenterForOpenScience/ember-osf/pull/110

## Purpose
Tracking PR; do not merge yet. Documents evolving requirements.

- Support cookie-based login
  - To support cookie origin rules, allow this app to be served from inside the OSF, eg `localhost:5000/preprints`. Requests are proxied to the underlying ember server.

- Update the application to use the newest version of ember, ember-cli, and ember-data (2.7).

## Summary of changes
This changes the application to use cookie auth.

This is not configurable yet, which means that you *must* run this branch within the OSF. Running under localhost:4200 will fail, because that port is not (usually) allowed to send cookie requests.

Also adds some namespacing to the mirage endpoints to avoid console errors.

### Changes required to run this code
This PR requires coordination with several other repositories.

1. ember-osf addon : include PRs 109 (merged) and [110](https://github.com/CenterForOpenScience/ember-osf/pull/110)
2. [OSF feature branch](https://github.com/CenterForOpenScience/osf.io/tree/feature/ember-preprints)
3. This branch of ember-preprints


You will also need:
- To reinstall all dependencies. This branch relies on upgrading ember version to Ember 2.7. (cli, ember, and ember-data)
- A running version of the OSF on your local computer. You are not allowed to use cookie auth with a remote server such as `test.osf.io`.
- If you were logging in with tokens, make sure ember-simple-auth doesn't mistakenly think it is logged in: at a JS console, run `localStorage.clear()`
- Make sure that your `config/local.yml` files specifies `REDIRECT_URI: http://localhost:5000/login/` for the local section. If you do not do this, the auth popup URL (from the login button) will say `service=undefined`

### Side effects
Some dependencies may need to be updated to work with new ember versions. Let me know if there are console error messages.

## How to test this
1. Log in (from any other OSF page under `localhost:5000`)
2. Navigate to `localhost:5000/preprints`. The navbar should show your OSF username and gravatar.
3. Go to any OSF page, and log out.
4. Go back to `localhost:5000/preprints`. The page should load and the navbar should show a login button.

## Eventually
Make the "login" button in the navbar work better (redirect to the desired ember page, relying on flask app to set cookie). For now it just sends the user to any other page. (For the reverse proxy making sure a cookie is set may be easier; just make sure set-cookie header comes along with the proxied response.)

Improve configuration, and remove any hard-coded references to `osf-token` in favor of configurable reference.

Don't break localhost:4200; let it fall back to oauth tokens if you want to develop without running the OSF locally.




## Ticket
https://openscience.atlassian.net/browse/PREP-87
